dask/tests/test_delayed.py

```diff
@@ -93,24 +94,132 @@ def test_delayed():
 
 
 def test_delayed_with_dataclass():
-    dataclasses = pytest.importorskip("dataclasses")
+    @dataclass()
+    class ADataClass:
+        a: int
 
-    # Avoid @dataclass decorator as Python < 3.7 fail to interpret the type hints
-    ADataClass = dataclasses.make_dataclass(
-        "ADataClass", [("a", int), ("b", int, dataclasses.field(init=False))]
-    )
+    literal = delayed(3)
+    with_class = delayed({"data": ADataClass(a=literal)})
+
+    def return_nested(obj):
+        return obj["data"].a
+
+    final = delayed(return_nested)(with_class)
+
+    assert final.compute() == 3
+
+
+def test_delayed_with_frozen_dataclass():
+    @dataclass(frozen=True)
+    class ADataClass:
+        a: int
+
+    literal = delayed(3)
+    with_class = delayed({"data": ADataClass(a=literal)})
+
+    def return_nested(obj):
+        return obj["data"].a
+
+    final = delayed(return_nested)(with_class)
+
+    assert final.compute() == 3
+
+
+def test_delayed_with_dataclass_with_custom_init():
+    @dataclass()
+    class ADataClass:
+        a: int
+
+        def __init__(self, b: int):
+            self.a = b
 
     literal = dask.delayed(3)
-    with_class = dask.delayed({"a": ADataClass(a=literal)})
+
+    with pytest.raises(TypeError) as e:
+        dask.delayed({"data": ADataClass(b=literal)})
+
+    e.match(r"ADataClass")
+    e.match(r"custom __init__ is not supported")
+
+
+def test_delayed_with_eager_dataclass_with_set_init_false_field():
+    @dataclass
+    class ADataClass:
+        a: int
+        b: int = field(init=False)
+
+    def prep_dataclass(a):
+        data = ADataClass(a=a)
+        data.b = 4
+        return data
+
+    with_class = delayed({"data": prep_dataclass(3)})
 
     def return_nested(obj):
-        return obj["a"].a
+        return obj["data"].a
 
     final = delayed(return_nested)(with_class)
 
     assert final.compute() == 3
 
 
+def test_delayed_with_dataclass_with_set_init_false_field():
+    @dataclass
+    class ADataClass:
+        a: int
+        b: int = field(init=False)
+
+    literal = dask.delayed(3)
+
+    def prep_dataclass(a):
+        data = ADataClass(a=a)
+        data.b = 4
+        return data
+
+    with pytest.raises(ValueError) as e:
+        dask.delayed(prep_dataclass(literal))
+
+    e.match(r"ADataClass")
+    e.match(r"`init=False` are not supported")
+
+
+def test_delayed_with_dataclass_with_unset_init_false_field():
+    @dataclass
+    class ADataClass:
+        a: int
+        b: int = field(init=False)
+
+    literal = dask.delayed(3)
+    with_class = delayed({"data": ADataClass(a=literal)})
+
+    def return_nested(obj):
+        return obj["data"].a
+
+    final = delayed(return_nested)(with_class)
+
+    assert final.compute() == 3
```

Comment:
Phew, trying to draw out the case where a `field(init=False)` field works is really tough. I must say, after reading through these three tests, I'm still pretty puzzled by it (which, I think, speaks to your desire to deprecate this code path overall).